src/components/InfiniteScrollPaginator/InfiniteScroll.js

```diff
@@ -1,232 +1,156 @@
+};
+
+/** @param {import("types").InfiniteScrollProps} props */
+// eslint-disable-next-line sonarjs/cognitive-complexity
+const InfiniteScroll = (props) => {
+  const {
+    children,
+    element = 'div',
+    hasMore = false,
+    initialLoad = true,
+    isReverse = false,
+    loader,
+    loadMore,
+    threshold = 250,
+    useCapture = false,
+    useWindow = true,
+    isLoading = false,
+    listenToScroll,
+    ...elementProps
+  } = props;
+  /** @type {React.MutableRefObject<HTMLElement | null>} */
+  const scrollComponent = useRef(null);
+
+  const scrollListener = useCallback(() => {
+    const el = scrollComponent.current;
+    if (!el) return;
+    const { parentElement } = el;
+
+    let offset = 0;
+    let reverseOffset = 0;
+    if (useWindow) {
       const doc =
         document.documentElement || document.body.parentNode || document.body;
       const scrollTop =
-        scrollEl.pageYOffset !== undefined
-          ? scrollEl.pageYOffset
-          : doc.scrollTop;
-      if (this.props.isReverse) {
-        offset = scrollTop;
-      } else {
-        offset = this.calculateOffset(el, scrollTop);
-      }
-    } else if (this.props.isReverse) {
-      offset = parentNode.scrollTop;
-    } else {
-      offset = el.scrollHeight - parentNode.scrollTop - parentNode.clientHeight;
+        window.pageYOffset !== undefined ? window.pageYOffset : doc.scrollTop;
+      offset = calculateOffset(el, scrollTop);
+      reverseOffset = scrollTop;
+    } else if (parentElement) {
+      offset =
+        el.scrollHeight - parentElement.scrollTop - parentElement.clientHeight;
+      reverseOffset = parentElement.scrollTop;
     }
+    if (listenToScroll) listenToScroll(offset, reverseOffset);
 
     // Here we make sure the element is visible as well as checking the offset
     if (
-      offset < Number(this.props.threshold) &&
-      el &&
-      el.offsetParent !== null
+      (isReverse ? reverseOffset : offset) < Number(threshold) &&
+      el.offsetParent !== null &&
+      typeof loadMore === 'function'
     ) {
-      this.detachScrollListener();
-      // Call loadMore after detachScrollListener to allow for non-async loadMore functions
-      if (typeof this.props.loadMore === 'function') {
-        this.props.loadMore();
-      }
+      loadMore();
     }
-  }
+  }, [useWindow, isReverse, threshold, listenToScroll, loadMore]);
 
-  calculateOffset(el, scrollTop) {
-    if (!el) {
-      return 0;
+  useEffect(() => {
+    const scrollEl = useWindow ? window : scrollComponent.current?.parentNode;
+    if (!hasMore || isLoading || !scrollEl) {
+      return () => {};
     }
 
-    return (
-      this.calculateTopPosition(el) +
-      (el.offsetHeight - scrollTop - window.innerHeight)
-    );
-  }
+    scrollEl.addEventListener('scroll', scrollListener, useCapture);
+    scrollEl.addEventListener('resize', scrollListener, useCapture);
 
-  calculateTopPosition(el) {
-    if (!el) {
-      return 0;
+    if (initialLoad) {
```

Comment:
Could we get rid of the `initialLoad` prop (by adding an extra `useEffect(scrollListener)`)? I can't think of a reason for exposing this functionality.